fedimint-client/src/oplog.rs

```diff
@@ -64,45 +67,107 @@ impl OperationLog {
         limit: usize,
         start_after: Option<ChronologicalOperationLogKey>,
     ) -> Vec<(ChronologicalOperationLogKey, OperationLogEntry)> {
+        const EPOCH_DURATION: Duration = Duration::from_secs(60 * 60 * 24 * 7);
+
         let mut dbtx = self.db.begin_transaction_nc().await;
-        let operations: Vec<ChronologicalOperationLogKey> = dbtx
-            .find_by_prefix_sorted_descending(&ChronologicalOperationLogKeyPrefix)
-            .await
-            .map(|(key, ())| key)
-            // FIXME: this is a schlemil-the-painter algorithm that will take longer the further
-            // back in history one goes. To avoid that I see two options:
-            //   1. Add a reference to the previous operation to each operation log entry,
-            //      essentially creating a linked list, which seem a little bit inelegant.
-            //   2. Add an option to prefix queries that allows to specify a start key
-            //
-            // The current implementation may also skip operations due to `SystemTime` not being
-            // guaranteed to be monotonous. The linked list approach would also fix that.
-            .skip_while(|key| {
-                let skip = if let Some(start_after) = start_after {
-                    key.creation_time >= start_after.creation_time
-                } else {
-                    false
-                };
-
-                std::future::ready(skip)
-            })
-            .take(limit)
-            .collect::<Vec<_>>()
-            .await;
+        let operation_log_keys = if let Some(start_after) = start_after {
+            let Some(last_key) = dbtx
+                .find_by_prefix(&ChronologicalOperationLogKeyPrefix)
+                .await
+                .map(|(key, ())| key)
+                .next()
+                .await
+            else {
+                return vec![];
+            };
+
+            // We want to fetch all operations that were created before `start_after`, going
+            // backwards in time. This means "start" generally means a later time than
+            // "end". Only when creating a rust Range we have to swap the terminology (see
+            // comment there).
+            let epochs_rev_ranges = (0..)
+                .map(|epoch| start_after.creation_time - epoch * EPOCH_DURATION)
+                // We want to get all operation log keys in the range [last_key, start_after). Sp as
+                // long as the start time is greater than the last key's creation time, we have to
+                // keep going.
+                .take_while(|&start_time| start_time >= last_key.creation_time)
```

Comment:
I thought that it will be simpler.

IIUC, we want to paginate from the newest to the oldest?

In that case I would just always query `find_by_range(last_seen-EPOCH..last_seen)`, and use `start_key` (minimum know in the collection), to just terminate pagination when `last_seen < start_key`.

If querying `find_by_range(last_seen-EPOCH..last_seen)` returns too many entries, just throw away extra ones. If not enough, loop and query for more, the concat.

Reply:
> IIUC, we want to paginate from the newest to the oldest (reverse order)?

:+1: 

> In that case I would just always query `find_by_range(last_seen-EPOCH..last_seen)`, and use `start_key` (minimum know in the collection), to just terminate pagination when `last_seen < start_key`.

That's what happens here, isn't it? I just don't want the caller of `list_operations` to worry about that end timestamp, but rather abort pagination when `None` is returned.

Reply:
We're using slower `find_by_prefix_sorted_descending` here and generally I'm confused about this function.

Pseudo-codishly:

```rust
let mut entries = vec![];
let mut last_seen = last_seen;

while log_entries.len() < limit {
  if last_seen < first_ever_event {
    // reached the beginning, no point in querying for more
    break;
  }
  let needs_num_entries = limit - entries.len();
  // get PERIOD worth of entries
  let more_entries = db.get_prefix(last_seen-PERIOD..last_seen);
  last_seen = more_entries.last();
  // take last `needs_num_entries`
  let more_entries = log_entries[more_entries.len().satuarating_sub(needs_num_entries)..more_entries.len()];
  log_entries = concat([more_events, entries]);
}

entries
```

Reply:
For the records: I was explained that I'm just confused and the existing code works kind of like it already.